# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [13]:
# Import your libraries:
from sklearn import datasets
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_diabetes.html)).

In [2]:
# Your code here:
diabetes = datasets.load_diabetes()

Let's explore this variable by looking at the different attributes (keys) of `diabetes`. Note that the `load_diabetes` function does not return dataframes. It returns you a Python dictionary.

In [3]:
# Your code here:
diabetes.keys()

dict_keys(['data', 'target', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

#### The next step is to read the description of the dataset. 

Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable. Read the data description carefully to fully understand what each column represents.

*Hint: If your output is ill-formatted by displaying linebreaks as `\n`, it means you are not using the `print` function.*

In [4]:
# Your code here:
# otra opcion --> diabetes.DESCR
print(diabetes["DESCR"])

.. _diabetes_dataset:

Diabetes dataset
----------------

Ten baseline variables, age, sex, body mass index, average blood
pressure, and six blood serum measurements were obtained for each of n =
442 diabetes patients, as well as the response of interest, a
quantitative measure of disease progression one year after baseline.

**Data Set Characteristics:**

  :Number of Instances: 442

  :Number of Attributes: First 10 columns are numeric predictive values

  :Target: Column 11 is a quantitative measure of disease progression one year after baseline

  :Attribute Information:
      - Age
      - Sex
      - Body mass index
      - Average blood pressure
      - S1
      - S2
      - S3
      - S4
      - S5
      - S6

Note: Each of these 10 feature variables have been mean centered and scaled by the standard deviation times `n_samples` (i.e. the sum of squares of each column totals 1).

Source URL:
https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html

For more information see:
Bra

#### Based on the data description, answer the following questions:

1. How many attributes are there in the data? What do they mean?

1. What is the relation between `diabetes['data']` and `diabetes['target']`?

1. How many records are there in the data?

### Enter your answer here:

1) Hay 10 atributos dentro de data: edad, sexo, indice de masa corporal, promedio de presión snaguínea y seis medidas de suero de 442 pacientes. Son valores predictivos del estadio de la enfermedad.

2) Para cada data (que es un datapoint con 10 atributos) o cada record se da un medida cuantitativa de la progresion de la enfermedad después de un año. Es el estadio de la enfermedad.

3) Hay 442 records

#### Now explore what are contained in the *data* portion as well as the *target* portion of `diabetes`. 

Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). Inspect the shape of `data` and `target`. Confirm they are consistent with the data description.

In [5]:
# Your code here:
display(diabetes["data"].shape,diabetes["target"].shape)

(442, 10)

(442,)

# Challenge 2 - Perform Supervised Learning on the Dataset

The data have already been split to predictor (*data*) and response (*target*) variables. Given this information, we'll apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset.

#### Let's briefly revisit the linear regression formula:

```
y = β0 + β1X1 + β2X2 + ... + βnXn + ϵ
```

...where:

- X1-Xn: data 
- β0: intercept 
- β1-βn: coefficients 
- ϵ: error (cannot explained by model)
- y: target

Also take a look at the `sklearn.linear_model.LinearRegression` [documentation](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html).

#### In the cell below, import the `linear_model` class from `sklearn`. 

In [6]:
# Your code here:
from sklearn.linear_model import LinearRegression

#### Create a new instance of the linear regression model and assign the new instance to the variable `diabetes_model`.

In [7]:
# Your code here:
diabetes_model = LinearRegression()

#### Next, let's split the training and test data.

Define `diabetes_data_train`, `diabetes_target_train`, `diabetes_data_test`, and `diabetes_target_test`. Use the last 20 records for the test data and the rest for the training data.

In [8]:
# Your code here:
diabetes_data_train = diabetes["data"][:422]
diabetes_target_train = diabetes["target"][:422]
diabetes_data_test = diabetes["data"][422:]
diabetes_target_test = diabetes["target"][422:]

Fit the training data and target to `diabetes_model`. Print the *intercept* and *coefficients* of the model.

In [9]:
# Your code here:
reg = diabetes_model.fit(diabetes_data_train,diabetes_target_train)
print(reg.intercept_)
print(reg.coef_)

152.76430691633442
[ 3.03499549e-01 -2.37639315e+02  5.10530605e+02  3.27736980e+02
 -8.14131709e+02  4.92814588e+02  1.02848452e+02  1.84606489e+02
  7.43519617e+02  7.60951722e+01]


In [10]:
# Son los coeficientes y el coeficiente de intercepción de la ecuación de regresión lineal que ha calculado el modelo

#### Inspecting the results

From the outputs you should have seen:

- The intercept is a float number.
- The coefficients are an array containing 10 float numbers.

This is the linear regression model fitted to your training dataset.

#### Using your fitted linear regression model, predict the *y* of `diabetes_data_test`.

In [11]:
# Your code here:
y_pred = reg.predict(diabetes_data_test)

#### Print your `diabetes_target_test` and compare with the prediction. 

In [12]:
# Your code here:
results = pd.DataFrame({"GT":diabetes_target_test,"prediction":y_pred})
results

GT  prediction
0   233.0  197.618469
1    91.0  155.439793
2   111.0  172.886651
3   152.0  111.535373
4   120.0  164.800548
5    67.0  131.069549
6   310.0  259.122378
7    94.0  100.479352
8   183.0  117.060105
9    66.0  124.305036
10  173.0  218.366328
11   72.0   61.198313
12   49.0  132.250468
13   64.0  120.333292
14   48.0   52.544587
15  178.0  194.037981
16  104.0  102.571397
17  132.0  123.566050
18  220.0  211.034632
19   57.0   52.603357

In [15]:
from sklearn.metrics import mean_squared_error

mean_squared_error(diabetes_target_test, y_pred)

2004.5676026898207

#### Is `diabetes_target_test` exactly the same as the model prediction? Explain.

In [17]:
# Your explanation here:
# El modelo no predice bien los valores del Ground truth. Además, el mse es muy alto, por tanto el modelo no es bueno  

# Bonus Challenge 1 - Hypothesis Testing with `statsmodels`

After generating the linear regression model from the dataset, you probably wonder: then what? What is the statistical way to know if my model is reliable or not?

Good question. We'll discuss that using Scikit-Learn in Challenge 5. But for now, let's use a fool-proof way by using the ([Linear Regression class of StatsModels](https://www.statsmodels.org/dev/regression.html)) which can also conduct linear regression analysis plus much more such as calcuating the F-score of the linear model as well as the standard errors and t-scores for each coefficient. The F-score and t-scores will tell you whether you can trust your linear model.

To understand the statistical meaning of conducting hypothesis testing (e.g. F-test, t-test) for slopes, read [this webpage](https://onlinecourses.science.psu.edu/stat501/node/297/) at your leisure time. We'll give you a brief overview next.

* The F-test of your linear model is to verify whether at least one of your coefficients is significantly different from zero. Translating that into the *null hypothesis* and *alternative hypothesis*, that is:

    ```
    H0 : β1 = β2 = ... = β10 = 0
    HA : At least one βj ≠ 0 (for j = 1, 2, ..., 10)
    ```

* The t-tests on each coefficient is to check whether the confidence interval for the variable contains zero. If the confidence interval contains zero, it means the null hypothesis for that variable is not rejected. In other words, this particular vaiable is not contributing to your linear model and you can remove it from your formula.

Read the documentations of [StatsModels Linear Regression](https://www.statsmodels.org/dev/regression.html) as well as its [`OLS` class](https://www.statsmodels.org/dev/generated/statsmodels.regression.linear_model.OLS.html) which stands for *ordinary least squares*.

#### In the next cell, analyze `diabetes_data_train` and `diabetes_target_train` with the linear regression model of `statsmodels`. Print the fit summary.

Your output should look like:

![statsmodels regression](../statsmodels.png)

In [18]:
# Your code here:
import statsmodels.api as sm
model = sm.OLS(diabetes_target_train,diabetes_data_train)
res = model.fit()
print(res.summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.110
Model:                            OLS   Adj. R-squared (uncentered):              0.089
Method:                 Least Squares   F-statistic:                              5.109
Date:                Tue, 05 May 2020   Prob (F-statistic):                    4.77e-07
Time:                        21:40:40   Log-Likelihood:                         -2745.5
No. Observations:                 422   AIC:                                      5511.
Df Residuals:                     412   BIC:                                      5552.
Df Model:                          10                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

#### Interpreting hypothesis testing results

Answer the following questions in the cell below:

1. What is the F-score of your linear model and is the null hypothesis rejected?

1. Does any of the t-tests of the coefficients produce a confidence interval containing zero? What are they?

1. How will you modify your linear reguression model according to the test results above?

### Your answers here:
1) F-statistic = 5.109 y el p-value es 4.77e-07. Este p-value es menor de 0,05, por lo que en un intervalo de confianza de 95%, podemos rechazar la hipotesis nula y por tanto decir que al menos uno de los coeficientes no es igual a cero, es decir, que al menor uno de los valores afecta a la variable dependiente.

2) P>|t| es menor de 0.05 para el coeficiente 3, lo que significa que no es igual a cero y por tanto sería el único que contribuye al modelo de regresión lineal. Y el coeficiente 4 tiene un p-value de 0.08 para un test two-tailed, por lo que para un one-tailed test sí que estaría por debajo de 0.05 y también lo podríamos considerar para el cálculo de la variable dependiente.

3) Por tanto, modificaría el modelo de regresión lineal dejando X (o el conjunto de datos que van a predecir) en las variables 3 y 4 que son body mass index y average blood pressure.

# Challenge 3 - Peform Supervised Learning on a Pandas Dataframe

Now that we have dealt with data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [19]:
# Your code here:
auto = pd.read_csv("../auto-mpg.csv")

Look at the first 5 rows using the `head()` function:

In [20]:
# Your code here:
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

In [21]:
# Your code here:
auto.dtypes

mpg             float64
cylinders         int64
displacement    float64
horse_power     float64
weight            int64
acceleration    float64
model_year        int64
car_name         object
dtype: object

What is the newest model year and the oldest model year?

In [22]:
# Your code here:
oldest = auto["model_year"].min()
newest = auto["model_year"].max()
print(f"The oldest model year is {oldest} and the newest {newest}")

The oldest model year is 70 and the newest 82


Check the dataset for missing values and remove all rows containing at least one missing value.

In [23]:
# Your code here:
auto.isnull().sum()

mpg             0
cylinders       0
displacement    0
horse_power     6
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

In [24]:
auto.dropna(axis=0,how="any",inplace=True)

In [25]:
auto.isnull().sum()

mpg             0
cylinders       0
displacement    0
horse_power     0
weight          0
acceleration    0
model_year      0
car_name        0
dtype: int64

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [26]:
# Your code here:
auto["cylinders"].value_counts()

4    199
8    103
6     83
3      4
5      3
Name: cylinders, dtype: int64

In [27]:
#Hay 5 posibles valores para cylinders

We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

*Hint: To separate data for training and test, use the `train_test_split` method we used in previous labs.*

In [28]:
# Your code here:
auto.drop(columns="car_name",inplace=True)

In [29]:
predictor = auto["mpg"]
response = auto.drop(columns="mpg")

In [30]:
X_train, X_test, y_train, y_test = train_test_split(response, predictor, test_size=0.2)

Now we will processed and peform linear regression on this data to predict the mpg for each vehicle. 

#### In the next cell, create an instance of the linear regression model and call it `auto_model`. Fit `auto_model` with your training data.

In [31]:
# Your code here:
auto_model=LinearRegression()
model = auto_model.fit(X_train,y_train)

# Challenge 4 - Evaluate the Model

In addition to evaluating your model with F-test and t-test, you can also use the *Coefficient of Determination* (a.k.a. *r squared score*). This method does not simply tell *yes* or *no* about the model fit but instead indicates how much variation can be explained by the model. Based on the r squared score, you can decide whether to improve your model in order to obtain a better fit.

You can learn about the r squared score [here](). Its formula is:

![R Squared](../r-squared.png)

...where:

* yi is an actual data point.
* ŷi is the corresponding data point on the estimated regression line.

By adding the squares of the difference between all yi-ŷi pairs, we have a measure called SSE (*error sum of squares*) which is an application of the r squared score to indicate the extent to which the estimated regression model is different from the actual data. And we attribute that difference to the random error that is unavoidable in the real world. Obviously, we want the SSE value to be as small as possible.

#### In the next cell, compute the predicted *y* based on `X_train` and call it `y_pred`. Then calcualte the r squared score between `y_pred` and `y_train` which indicates how well the estimated regression model fits the training data.

*Hint: r squared score can be calculated using `sklearn.metrics.r2_score` ([documentation](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html)).*

In [32]:
# Your code here:
y_pred = model.predict(X_train)

In [33]:
from sklearn.metrics import r2_score
r2_score(y_train, y_pred)

0.8101849711531378

#### Our next step is to evaluate the model using the test data. 

We would like to ensure that our model is not overfitting the data. This means that our model was made to fit too closely to the training data by being overly complex. If a model is overfitted, it is not generalizable to data outside the training data. In that case, we need to reduce the complexity of the model by removing certain features (variables).

In the cell below, use the model to generate the predicted values for the test data and assign them to `y_test_pred`. Compute the r squared score of the predicted `y_test_pred` and the oberserved `y_test` data.

In [34]:
# Your code here:
y_test_pred = model.predict(X_test)
r2_score(y_test_pred, y_test)

0.7149049106819779

#### Explaining the results

The r squared scores of the training data and the test data are pretty close (0.8146 vs 0.7818). This means our model is not overfitted. However, there is still room to improve the model fit. Move on to the next challenge.

# Challenge 5 - Improve the Model Fit

While the most common way to improve the fit of a model is by using [regularization](https://datanice.github.io/machine-learning-101-what-is-regularization-interactive.html), there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [35]:
# Your code here:
X_train09, X_test09, y_train09, y_test09= train_test_split(response, predictor, test_size=0.1)

Initialize a new linear regression model. Name this model `auto_model09`. Fit the model to the new sample (training) data.

In [36]:
# Your code here:
auto_model09 = LinearRegression()
model = auto_model.fit(X_train09,y_train09)

Compute the predicted values and r squared score for our new model and new sample data.

In [37]:
# Your code here:
y_pred09 = model.predict(X_train09)
r2_score(y_train09, y_pred09)

0.8168562207298466

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [38]:
# Your code here:
y_test_pred09 = model.predict(X_test09)
r2_score(y_test_pred09, y_test09)

0.5852928049747415

In [39]:
# Ha empeorado. Aunque como el split es aleatorio irá variando conforme haga nuevos intentos

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

In [40]:
from sklearn.feature_selection import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

In [41]:
# Your code here:
auto_model=LinearRegression()
selector = RFE(auto_model, n_features_to_select=3, step=1, verbose=0)

Fit the model and print the ranking

In [42]:
# Your code here:
selector = selector.fit(X_train, y_train)
selector.ranking_

array([1, 2, 4, 3, 1, 1])

- cylinders       1
- displacement    4
- horse_power     2
- weight          3
- acceleration    1
- model_year      1

Es decir, cilindros, aceleracion y model_year son los 3 atributos que más afectan a la prediccion de mpg

Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [49]:
# Your code here:
X = auto.drop(columns=["displacement","horse_power","weight","mpg"])

X_train_reduced, X_test_reduced, y_train_reduced, y_test_reduced= train_test_split(X, predictor, test_size=0.2)

In [50]:
X.columns

Index(['cylinders', 'acceleration', 'model_year'], dtype='object')

Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [51]:
# Your code here: 
auto_model_reduced=LinearRegression()
model = auto_model_reduced.fit(X_train_reduced,y_train_reduced)

In [52]:
y_pred_red = model.predict(X_train_reduced)
print(r2_score(y_train_reduced, y_pred_red))
y_prdtest_red = model.predict(X_test_reduced)
print(r2_score(y_test_reduced, y_prdtest_red))

0.7166314008948269
0.7003348471967596


In [53]:
# Ha mejorado el R cuadrado

#### Conclusion

You may obtain the impression from this lab that without knowing statistical methods in depth, it is difficult to make major progress in machine learning. That is correct. If you are motivated to become a data scientist, statistics is the subject you must be proficient in and there is no shortcut. 

Completing these labs is not likely to make you a data scientist. But you will have a good sense about what are there in machine learning and what are good for you. In your future career, you can choose one of the three tracks:

* Data scientists who need to be proficient in statistical methods.

* Data engineers who need to be good at programming.

* Data integration specialists who are business or content experts but also understand data and programming. This cross-disciplinary track brings together data, technology, and business and will be in high demands in the next decade.